src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -199,33 +199,65 @@ public SNITCPHandle(string serverName, int port, long timerExpire, object callba
             }
 
-            // Try each address in turn, and return the socket opened for the first one that works.
-            ExceptionDispatchInfo lastException = null;
+            // Connect to each IP address in parallel, and return the socket opened for the first one that works.
+            List<Task> connectTasks = new List<Task>(serverAddresses.Length);
+
+            object socketOpenSetResultLock = new object();
+            Socket resultSocket = null;
+            ExceptionDispatchInfo lastSocketConnectEx = null;
+
+            CancellationTokenSource cancelTokenSource = new CancellationTokenSource();
+            CancellationToken cancelToken = cancelTokenSource.Token;
+
             foreach (IPAddress address in serverAddresses)
             {
-                var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+                Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
                 try
                 {
-                    await socket.ConnectAsync(address, port).ConfigureAwait(false);
-                    return socket;
+                    Task socketTask = socket.ConnectAsync(address, port).ContinueWith(
+                        t =>
+                        {
+                            lock (socketOpenSetResultLock)
+                            {
+                                if (!cancelToken.IsCancellationRequested && socket.Connected)
+                                {
+                                    // Successful socket connection, so cancel other pending tasks
+                                    resultSocket = socket;
+                                    cancelTokenSource.Cancel();
+                                    return;
+                                }
+                            }
+
+                            // Didn't return from setting result, so dispose this socket
+                            socket.Dispose();
+                        },
+                        cancelToken);
+
+                    connectTasks.Add(socketTask);
                 }
                 catch (Exception exc)
                 {
                     socket.Dispose();
-                    lastException = ExceptionDispatchInfo.Capture(exc);
+                    lastSocketConnectEx = ExceptionDispatchInfo.Capture(exc);
                 }
             }
 
-            // Propagate the last failure that occurred
-            if (lastException != null)
+            try
+            {
+                await Task.WhenAll(connectTasks).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException)
+            {
+                // Expected. When a connect task completes successfully, other pending tasks will be cancelled
+            }
+
+            if (lastSocketConnectEx != null && resultSocket == null)
             {
-                lastException.Throw();
+                lastSocketConnectEx.Throw();
```

Comment:
The tasks being stored into the list are the continuation tasks, not the original ConnectAsync-returned tasks, and nothing in the continuation is propagating exceptions that may have occurred on the ConnectAsync task.  So if all of the connections fail, won't lastSocketConnectEx be null and resultSocket will be null, and we'll end up returning null rather than throwing one of the exceptions that occurred?  I guess that's desired, as I see above at the call site you check for null and throw a different exception in that case?